### PR TITLE
Refactoring WANDPowderReduction algorithm to allow group workspaces as output

### PR DESCRIPTION
**Pull request by jenest**
**Description of work.**

Refactored code to fix bug when the user wants multiple input workspaces and a group workspace as output.

Fixes GitLab issue [PD#109](https://code.ornl.gov/sns-hfir-scse/diffraction/powder/powder-diffraction/-/issues/109)
Version into [ornl-next branch](https://github.com/mantidproject/mantid/pull/29990)

**To test:**

From CLI:./mantidpython /path/to/WANDPowderReductionTest.py

From MantidWorkbench editor: 
_note: this is included as a unit test in WANDPowderReductionTest.py_
```
from mantid.simpleapi import (
    WANDPowderReduction,
    CreateSampleWorkspace,
)
from mantid.api import WorkspaceGroup
import numpy as np


event_data = CreateSampleWorkspace(
    NumBanks=1,
    BinWidth=20000,
    PixelSpacing=0.1,
    BankPixelWidth=100,
    WorkspaceType="Event",
)
event_cal = CreateSampleWorkspace(
    NumBanks=1,
    BinWidth=20000,
    PixelSpacing=0.1,
    BankPixelWidth=100,
    WorkspaceType="Event",
    Function="Flat background",
)
event_bkg = CreateSampleWorkspace(
    NumBanks=1,
    BinWidth=20000,
    PixelSpacing=0.1,
    BankPixelWidth=100,
    WorkspaceType="Event",
    Function="Flat background",
)

pd_out = WANDPowderReduction(
    InputWorkspace=[event_data, event_data],
    CalibrationWorkspace=event_cal,
    BackgroundWorkspace=event_bkg,
    Target="Theta",
    NumberBins=1000,
    NormaliseBy="None",
    Sum=False,
)

for i in pd_out:
    x = i.extractX()
    y = i.extractY()
    assert x.min() == 0.03517355089700467
    assert x.max() == 70.3119282431118

assert isinstance(pd_out, WorkspaceGroup)
assert len(pd_out) == 2
```

Fixes [PD#109](https://code.ornl.gov/sns-hfir-scse/diffraction/powder/powder-diffraction/-/issues/109)
Version into [ornl-next branch](https://github.com/mantidproject/mantid/pull/29990)

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by rosswhitfield — CHANGES_REQUESTED** [review]
When I run your example I end up with two workspaces `event_bkgpd_out1` and `event_bkgpd_out2` that I assume are meant to be temporary and should be removed at some point in the algorithm.

In your example, if I change to `Sum=True` I end up getting the output twice as `pd_out` and `pd_out1`.


Should be this able to accept a workspace group as input? If I extend your example with
```python
event_data2 = CloneWorkspace(event_data)
event_data_group = GroupWorkspaces([event_data,event_data2])
pd_out = WANDPowderReduction(
    InputWorkspace=event_data_group,
    CalibrationWorkspace=event_cal,
    BackgroundWorkspace=event_bkg,
    Target="Theta",
    NumberBins=1000,
    NormaliseBy="None",
    Sum=False,
)
```

it fails with
```
WANDPowderReduction-[Notice] WANDPowderReduction started
WANDPowderReduction-[Error] Error in execution of algorithm WANDPowderReduction:
WANDPowderReduction-[Error] Execution of Transpose for group entry 1 failed: Transpose: error (see log)
WANDPowderReduction-[Error]   at line 153 in '/home/rwp/src/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/WANDPowderReduction.py'
WANDPowderReduction-[Error]   caused by line 341 in '/home/rwp/src/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/WANDPowderReduction.py'
WANDPowderReduction-[Error]   caused by line 292 in '/home/rwp/src/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/WANDPowderReduction.py'
WANDPowderReduction-[Error]   caused by line 1120 in '/home/rwp/src/mantid/Framework/PythonInterface/mantid/simpleapi.py'
RuntimeError: Execution of Transpose for group entry 1 failed: Transpose: error (see log)
  at line 153 in '/home/rwp/src/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/WANDPowderReduction.py'
  caused by line 341 in '/home/rwp/src/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/WANDPowderReduction.py'
  caused by line 292 in '/home/rwp/src/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/WANDPowderReduction.py'
  caused by line 1120 in '/home/rwp/src/mantid/Framework/PythonInterface/mantid/simpleapi.py'
  File "/home/rwp/test.py", line 1, in <module>
    # import mantid algorithms, numpy and matplotlib
  File "/home/rwp/src/mantid/Framework/PythonInterface/mantid/simpleapi.py", line 1120, in __call__
    algm.execute()
```

which appear to be the same error message as seen in  [PD#109](https://code.ornl.gov/sns-hfir-scse/diffraction/powder/powder-diffraction/-/issues/109)

**Comment by jenest** [comment]
> When I run your example I end up with two workspaces `event_bkgpd_out1` and `event_bkgpd_out2` that I assume are meant to be temporary and should be removed at some point in the algorithm.
> 
> In your example, if I change to `Sum=True` I end up getting the output twice as `pd_out` and `pd_out1`.
> 
> Should be this able to accept a workspace group as input? If I extend your example with
> 
> ```python
> event_data2 = CloneWorkspace(event_data)
> event_data_group = GroupWorkspaces([event_data,event_data2])
> pd_out = WANDPowderReduction(
>     InputWorkspace=event_data_group,
>     CalibrationWorkspace=event_cal,
>     BackgroundWorkspace=event_bkg,
>     Target="Theta",
>     NumberBins=1000,
>     NormaliseBy="None",
>     Sum=False,
> )
> ```
> 

Extra temporary workspaces should be taken care of now. 

Output is no longer duplicated.

It should work with a workspace group as input specified like the following (which is added to WANDPowderReductionTest.py):
```
event_data2 = CloneWorkspace(event_data)
group = WorkspaceGroup()
group.addWorkspace(event_data)
group.addWorkspace(event_data2)

pd_out = WANDPowderReduction(
    InputWorkspace=group,
    CalibrationWorkspace=event_cal,
    BackgroundWorkspace=event_bkg,
    Target="Theta",
    NumberBins=1000,
    NormaliseBy="None",
    Sum=False,
)
for i in pd_out:
    x = i.extractX()
    y = i.extractY()

    assert x.min() == 0.03517355089700467
    assert x.max() == 70.3119282431118

assert isinstance(pd_out, WorkspaceGroup)
assert len(pd_out) == 2
```

**Comment by KedoKudo** [comment]
* Run the unittest module in Workbench env on RHEL_7, and all tests passed.
* Other than the following issue, everything else looks good

__The individual reduce results seem to be affected when input are a combined workspace group.  To reproduce the issue, try the following code on analysis cluster__
```python
# load individual data set
si1 = LoadWAND('/HFIR/HB2C/IPTS-22745/nexus/HB2C_320234.nxs.h5')
si2 = LoadWAND('/HFIR/HB2C/IPTS-22745/nexus/HB2C_320235.nxs.h5')
va0 = LoadWAND('/HFIR/HB2C/IPTS-23858/nexus/HB2C_320259.nxs.h5')

# reduce each workspace independently and then calculate the merged results
si1_reduced = WANDPowderReduction(InputWorkspace=si1, CalibrationWorkspace=va0, Target='Theta', NumberBins=500, NormaliseBy="Time")
si2_reduced = WANDPowderReduction(InputWorkspace=si2, CalibrationWorkspace=va0, Target='Theta', NumberBins=500, NormaliseBy="Time")
sin_reduced = WANDPowderReduction(InputWorkspace=[si1, si2], CalibrationWorkspace=va0, Target='Theta', NumberBins=500, NormaliseBy="Time", Sum=True) 
```
which will results in the following image
WAND_newReducedRstMergedDataset
Now if we combine the two workspace into a workspace group and try the reduction and sum, we get something different
```python
# Continue from above code
si_group = WorkspaceGroup()
si_group.addWorkspace(si1)
si_group.addWorkspace(si2)
# reduce each workspace independently
si_group_reduced_nosum = WANDPowderReduction(InputWorkspace=si_group, CalibrationWorkspace=va0, Target="Theta", NumberBins=500, NormaliseBy="Time", Sum=False)
# reduce as a merged one
si_group_reduced_sum = WANDPowderReduction(InputWorkspace=si_group, CalibrationWorkspace=va0, Target="Theta", NumberBins=500, NormaliseBy="Time", Sum=True)
```
which yields
WAND_newReducedRstMergedDataset_useGroup
Notice how the orange curve is much shorter when reduced in a workspace group.  This is most likely due to the fact that a global binning range is used when processed as a workspace group, which affects the results of individual workspace in a group.  
@jenest Could you confirm with the CIS which one of theresults should be consider the accepted one and adjust the code accordingly?

**Review by KedoKudo — CHANGES_REQUESTED** [review]

File: Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/WANDPowderReductionTest.py [hunk 1]
```diff
@@ -382,6 +394,65 @@ def test_event(self):
...
+        for i in pd_out:
+            x = i.extractX()
+            y = i.extractY()
+
+            self.assertAlmostEqual(x.min(), 0.03517355)
+            self.assertAlmostEqual(x.max(), 70.3119282)
+            self.assertAlmostEqual(y[0, 0], 0.0)
+
+        assert isinstance(pd_out, WorkspaceGroup)
+        assert len(pd_out) == 2
```
[the remark this answers is not in the document]
Comment (jenest) [reply Framework/PythonInterface/test/python/plugins/algorithms/WorkflowAlgorithms/WANDPowderReductionTest.py:449]: No; thank you for catching that! Case 3's InputWorkspace should be passed in as a list of workspaces (fixed now).

**Review by KedoKudo — DISMISSED** [review]

**Review by rosswhitfield — CHANGES_REQUESTED** [review]
So this still fails when running original script from PD109

I think the issue is with the WorkspaceGroup, when you created the test with `WorkspaceGroup.addWorkspace()` the workspace group is not added to the ADS, if you use the `GroupWorkspaces` algorithm it will be added. I don't understand why that makes a difference but this does fail when it shouldn't.

```python
event_data2 = CloneWorkspace(event_data)
event_data_group = GroupWorkspaces([event_data,event_data2])
pd_out = WANDPowderReduction(
    InputWorkspace=event_data_group,
    CalibrationWorkspace=event_cal,
    BackgroundWorkspace=event_bkg,
    Target="Theta",
    NumberBins=1000,
    NormaliseBy="None",
    Sum=False,
)
```

**Review by rosswhitfield — APPROVED** [review]

**Pull request merged**